src/variety/tents.js

```diff
@@ -0,0 +1,499 @@
+					if(this.notInputted) {
+						this.inputqnum();
+					}
+				}
+			}
+		},
+
+		inputcell_tents: function(value) {
+			var cell = this.getcell();
+			if (cell.isnull || cell === this.mouseCell || cell.qnum !== -1) {
+				return;
+			}
+			if (this.inputData === null) {
+				var current = cell.anum===2 ? 1 : cell.qsub===1 ? 2 : 0;
+
+				if(!!value) {
+					this.inputData = current !== value ? value : 0;
+				} else if(this.puzzle.getConfig("use")===1){
+					var next = this.btn === "left" ? 1 : 2;
+					this.inputData = current !== next ? next : 0;
+				} else {
+					var next = current + (this.btn === "left" ? 1 : -1);
+					this.inputData = (next+3)%3;
+				}
+			}
+
+			cell.setAnum(this.inputData===1 ? 2 : -1);
+			cell.setQsub(this.inputData===2 ? 1 : 0);
+			cell.draw();
+
+			this.mouseCell = cell;
+
+			if (this.inputData === 1) {
+				this.mousereset();
+			}
+		},
+
+		inputqnum_excell: function() {
+			var excell = this.getcell_excell();
+			if (excell.isnull || excell.group !== "excell") {
+				return;
+			}
+
+			if (excell !== this.cursor.getex()) {
+				this.setcursor(excell);
+			} else {
+				this.inputqnum_main(excell);
+			}
+		},
+
+		inputShade: function() {
+			this.inputFixedNumber(3);
+		},
+
+		inputDot: function() {
+			this.inputcell_tents(2);
+		}
+	},
+
+	KeyEvent: {
+		enablemake: true,
+
+		keyinput: function(ca) {
+			this.key_inputexcell(ca);
+			this.key_inputqnum(ca);
+		},
+		key_inputexcell: function(ca) {
+			var excell = this.cursor.getex(),
+				qn = excell.qnum;
+			var max = excell.getmaxnum();
+
+			if ("0" <= ca && ca <= "9") {
+				var num = +ca;
+
+				if (qn <= 0 || this.prev !== excell) {
+					if (num <= max) {
+						excell.setQnum(num);
+					}
+				} else {
+					if (qn * 10 + num <= max) {
+						excell.setQnum(qn * 10 + num);
+					} else if (num <= max) {
+						excell.setQnum(num);
+					}
+				}
+			} else if (ca === " " || ca === "-") {
+				excell.setQnum(-1);
+			} else {
+				return;
+			}
+
+			this.prev = excell;
+			this.cursor.draw();
+		}
+	},
+
+	TargetCursor: {
+		initCursor: function() {
+			this.init(-1, -1);
+		}
+	},
+
+	ExCell: {
+		disInputHatena: true,
+
+		maxnum: function() {
+			var bx = this.bx,
+				by = this.by;
+			if (bx === -1 && by === -1) {
+				return 0;
+			}
+			return by === -1 ? this.board.rows : this.board.cols;
+		},
+		minnum: 0
+	},
+
+	Cell: {
+		numberAsObject: true,
+		disInputHatena: true,
+		maxnum: 3,
+
+		posthook: {
+			qnum: function(num) { 
+				if (num >= 0) { this.setQsub(0);}
+			}
+		}
+	},
+
+	Board: {
+		hasborder: 1,
+		hasexcell: 1,
+
+		cols: 8,
+		rows: 8,
+
+		addExtraInfo: function() {
+			this.campgraph = this.addInfoList(this.klass.AreaCampGraph);
+		}
+	},
+
+	BoardExec: {
+		adjustBoardData: function(key, d) {
+			this.adjustEdge_1(key, d);
+		},
+		adjustBoardData2: function(key, d) {
+			this.adjustEdge_2(key, d);
+		}
+	},
+
+	// The rules state that there must be at least one valid matching between trees and tents.
```

Comment:
It's not clear to me that these are *the* rules for tents and trees. E.g., http://wiki.logic-masters.de/index.php?title=Zeltlager/de require an explicit matching, which would mean a 3x3 puzzle with four tents on the sides is ambiguous.

Not saying that the approach you're taking here is wrong, just that I'd still like to consider what to do about this.

Regardless of the rule interpretation, it would seem to be useful to be able to mark tree-tent connections.